controllers/replace_failed_process_groups_test.go

```diff
@@ -63,6 +69,595 @@ var _ = Describe("replace_failed_process_groups", func() {
+			Expect(targetProcessGroupStatus.GetCondition(fdbv1beta2.NodeTaintReplacing).ProcessGroupConditionType).To(Equal(fdbv1beta2.NodeTaintReplacing))
+
+			// cluster won't replace a failed process until GetTaintReplacementTimeSeconds() later
+			Expect(replaceFailedProcessGroups{}.reconcile(ctx.TODO(), clusterReconciler, cluster)).To(BeNil())
+			Expect(getRemovedProcessGroupIDs(cluster)).To(Equal([]fdbv1beta2.ProcessGroupID{}))
+
+			Eventually(func() *requeue {
+				result = replaceFailedProcessGroups{}.reconcile(ctx.TODO(), clusterReconciler, cluster)
+				return result
+			}).WithTimeout(time.Duration(cluster.GetTaintReplacementTimeSeconds()*3) * time.Second).WithPolling(1 * time.Second).ShouldNot(BeNil())
+
+			Expect(getRemovedProcessGroupIDs(cluster)).To(Equal([]fdbv1beta2.ProcessGroupID{targetPodProcessGroupID}))
+		})
+
+		It("should replace a pod that has NodeTaintReplacing condition but no longer has NodeTaintDetected condition", func() {
+			node.Spec.Taints = []corev1.Taint{
+				{
+					Key:       taintKeyMaintenance,
+					Value:     "rack_maintenance",
+					Effect:    corev1.TaintEffectNoExecute,
+					TimeAdded: &metav1.Time{Time: time.Now().Add(-time.Second * time.Duration(taintKeyMaintenanceDuration+1))},
+				},
+			}
+			log.Info("Taint node", "Node name", podOnTaintedNode.Name, "Node taints", node.Spec.Taints, "TaintTime", node.Spec.Taints[0].TimeAdded.Time, "Now", time.Now())
+			Expect(k8sClient.Update(ctx.TODO(), node)).NotTo(HaveOccurred())
+
+			processGroupsStatus, err := validateProcessGroups(ctx.TODO(), clusterReconciler, cluster, &cluster.Status, processMap, configMap, allPods, allPvcs, logger)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(len(processGroupsStatus)).To(BeNumerically(">", 4))
+			targetProcessGroupStatus := fdbv1beta2.FindProcessGroupByID(cluster.Status.ProcessGroups, targetPodProcessGroupID)
+			Expect(len(targetProcessGroupStatus.ProcessGroupConditions)).To(Equal(2))
+			Expect(targetProcessGroupStatus.GetCondition(fdbv1beta2.NodeTaintDetected).ProcessGroupConditionType).To(Equal(fdbv1beta2.NodeTaintDetected))
+			Expect(targetProcessGroupStatus.GetCondition(fdbv1beta2.NodeTaintReplacing).ProcessGroupConditionType).To(Equal(fdbv1beta2.NodeTaintReplacing))
+
+			Eventually(func() *requeue {
+				return replaceFailedProcessGroups{}.reconcile(ctx.TODO(), clusterReconciler, cluster)
+			}).WithTimeout(time.Duration(cluster.GetTaintReplacementTimeSeconds()*3) * time.Second).WithPolling(1 * time.Second).ShouldNot(BeNil())
+
+			Expect(getRemovedProcessGroupIDs(cluster)).To(Equal([]fdbv1beta2.ProcessGroupID{targetPodProcessGroupID}))
+		})
+
+		It("should not replace a pod that is on a flapping tainted node", func() {
+			// Flapping tainted node
+			tainted := int64(0)
+			for taintTimeOffset := taintKeyMaintenanceDuration * 2; taintTimeOffset >= 0; taintTimeOffset-- {
```

Comment:
I'm not sure I understand the idea behind this for loop, why do we need this? What is the benefit of running this for loop and updating the node object?

Reply:
I'm trying to create a constant flappying tainted node situation and test operator won't mark Pods on the flapping Node as TaintReplacing condition.